fhir_pyrate/ahoy.py

```diff
@@ -139,8 +142,25 @@ def refresh_token(self, token: str = None) -> None:
             self.token = response.text
             self.session.headers.update({"Authorization": f"Bearer {self.token}"})
             self.auth_time = datetime.datetime.now()
-        else:
-            logger.warning(
-                "The token cannot be refreshed because neither a valid token nor a refresh url has "
-                "been provided."
+        elif self.auth_type is not None and self.auth_method is not None:
+            self._authenticate()
+            self.auth_time = datetime.datetime.now()
+
+    def _refresh_hook(
+        self, response: requests.Response, *args: Any, **kwargs: Any
+    ) -> Optional[requests.Response]:
+        if self.login_reattempted_times >= self.max_login_attempts:
+            response.raise_for_status()
+        elif response.status_code == requests.codes.unauthorized:
+            self.login_reattempted_times += 1
```

Comment:
I would move this counter to the `response.request` object.

Reply:
in the header?

Reply:
After a quick search I did not find a "userdata" member variable, where you could store such information. However, you could just do `response.request.login_reattemted_times += 1` and use `hasattr(response.request, "login_reattemted_times")`.